### PR TITLE
Keepalive logic in zmq-cpiface

### PR DESCRIPTION
**Pull request by ajamshed**
Cleaned up the code.
Also, made sure that cpiface self-restarts for non-k8s setup as well.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -36,7 +36,6 @@ docker-build:
 		DOCKER_BUILDKIT=$(DOCKER_BUILDKIT) docker build --pull $(DOCKER_BUILD_ARGS) \
 			--target $$target \
 			--tag ${DOCKER_REGISTRY}${DOCKER_REPOSITORY}upf-epc-$$target:${DOCKER_TAG} \
-			--cache-from ${DOCKER_REGISTRY}${DOCKER_REPOSITORY}upf-epc-$$target:${DOCKER_TAG} \
```
Comment (ajamshed) [inline Makefile:39]: @krsna1729,

See if this deletion makes sense.

Comment (krsna1729) [reply Makefile:39]: also delete 

--build-arg BUILDKIT_INLINE_CACHE
BUILDKIT_INLINE_CACHE    ?= 1

**Comment by ajamshed** [comment]
This PR is not ready yet. Sessions need to be deleted before restart.

File: cpiface/zmq-cpiface.cc [hunk 2]
```diff
@@ -142,17 +146,56 @@ void sig_handler(int signo) {
...
+  pid = fork();
+
+  if (pid == -1) {
+    std::cerr << "Failed to fork: " << strerror(errno) << std::endl;
+    exit(EXIT_FAILURE);
+  } else if (pid == 0) {  // child process
+    execv(argv[0], argv);
+    exit(EXIT_SUCCESS);
+  } else {  // parent process
+    if (waitpid(pid, &status, 0) > 0) {
```
Comment (thakurajayL) [inline cpiface/zmq-cpiface.cc:162]: So child is new cpiface and parent (old cpiface) dies eventually. Is that right ?

Comment (ajamshed) [reply cpiface/zmq-cpiface.cc:162]: Yes.

File: Makefile [hunk 3]
```diff
@@ -18,8 +18,6 @@ DOCKER_REPOSITORY        ?=
 DOCKER_TAG               ?= ${VERSION}
 DOCKER_IMAGENAME         := ${DOCKER_REGISTRY}${DOCKER_REPOSITORY}${PROJECT_NAME}:${DOCKER_TAG}
 DOCKER_BUILDKIT          ?= 1
-BUILDKIT_INLINE_CACHE    ?= 1
-DOCKER_BUILD_ARGS        ?= --build-arg MAKEFLAGS=-j$(shell nproc) --build-arg CPU --build-arg BUILDKIT_INLINE_CACHE
```
Comment (krsna1729) [inline Makefile:22]: @ajamshed do not remove the entire line. Only `--build-arg BUILDKIT_INLINE_CACHE`

**Comment by krsna1729** [comment]
@ajamshed check the jenkins failure

**Comment by ajamshed** [comment]
Retest this please.

**Comment by thakurajayL** [comment]
Should we have similar code i case pfcp endpoint at (upf) detects Central Plane down ?

**Comment by ajamshed** [comment]
> Should we have similar code i case pfcp endpoint at (upf) detects Central Plane down ?

We should first rebase. Then we will need to add keepalive/heartbeat logic to both zmq-cpiface.cc as well as pfcpiface.go.

**Comment by thakurajayL** [comment]
> > Should we have similar code i case pfcp endpoint at (upf) detects Central Plane down ?
> 
> We should first rebase. Then we will need to add keepalive/heartbeat logic to both zmq-cpiface.cc as well as pfcpiface.go.

I am not sure. Let's forget about other things. This fix is proposed for master branch as of now. Do i have to do any action ? I dont see Git asking me to do rebase.

**Comment by ajamshed** [comment]
> > > Should we have similar code i case pfcp endpoint at (upf) detects Central Plane down ?
> > 
> > 
> > We should first rebase. Then we will need to add keepalive/heartbeat logic to both zmq-cpiface.cc as well as pfcpiface.go.
> 
> I am not sure. Let's forget about other things. This fix is proposed for master branch as of now. Do i have to do any action ? I dont see Git asking me to do rebase.

No other action needed. The PR should be merged to the master once the Jenkins test passes. I have already sent a message to @yoooou for help. Merging similar changes to the topics/up4-bess branch will require additional work since we need to clear more table entries (PDRLookup, FARLookup, Counters) before the upf restarts.

**Comment by yoooou** [comment]
retest this please

**Comment by yoooou** [comment]
retest this please

**Pull request merged**